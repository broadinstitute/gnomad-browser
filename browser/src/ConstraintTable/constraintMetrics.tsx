import React from 'react'
import styled from 'styled-components'

import { TooltipAnchor, TooltipHint } from '@gnomad/ui'

const ConstraintHighlight = styled.span`
  display: inline-block;
  padding: 0.25em 0.4em;
  border: 1px solid #000;
  border-radius: 0.3em;
  background: ${(props: any) => props.highlightColor};
  color: #000;
`

export const renderRoundedNumber = (
  num: number | null,
  {
    precision = 1 as number,
    tooltipPrecision = 3,
    highlightColor = null as string | null,
    formatTooltip = ((n) => `${n}`) as (n: number) => string,
  } = {}
) => {
  if (num === null) {
    return 'â€”'
  }

  const roundedNumber = Number(num.toFixed(precision)).toString()
  return (
    // @ts-expect-error TS(2322) FIXME: Type '{ children: Element; tooltip: any; }' is not... Remove this comment to see the full error message
    <TooltipAnchor tooltip={formatTooltip(num.toFixed(tooltipPrecision))}>
      {highlightColor ? (
        // @ts-expect-error TS(2769) FIXME: No overload matches this call.
        <ConstraintHighlight highlightColor={highlightColor}>{roundedNumber}</ConstraintHighlight>
      ) : (
        // @ts-expect-error TS(2745) FIXME: This JSX tag's 'children' prop expects type 'never... Remove this comment to see the full error message
        <TooltipHint>{roundedNumber}</TooltipHint>
      )}
    </TooltipAnchor>
  )
}
