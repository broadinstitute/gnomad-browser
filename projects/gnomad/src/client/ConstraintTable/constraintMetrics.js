import React from 'react'
import styled from 'styled-components'

import { TooltipAnchor, TooltipHint } from '@broad/ui'

const ConstraintHighlight = styled.span`
  display: inline-block;
  padding: 0.25em 0.4em;
  border: 1px solid #000;
  border-radius: 0.3em;
  background: ${props => props.highlightColor};
  color: #000;
`

export const renderRoundedNumber = (
  num,
  { precision = 1, tooltipPrecision = 3, highlightColor = null, formatTooltip = n => n } = {}
) => {
  if (num === null) {
    return 'â€”'
  }

  const roundedNumber = Number(num.toFixed(precision)).toString()
  return (
    <TooltipAnchor tooltip={formatTooltip(num.toFixed(tooltipPrecision))}>
      {highlightColor ? (
        <ConstraintHighlight highlightColor={highlightColor}>{roundedNumber}</ConstraintHighlight>
      ) : (
        <TooltipHint>{roundedNumber}</TooltipHint>
      )}
    </TooltipAnchor>
  )
}
