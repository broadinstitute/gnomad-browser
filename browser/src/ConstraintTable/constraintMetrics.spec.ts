import { expect, describe, test } from '@jest/globals'
import { renderRoundedNumber } from './constraintMetrics'

describe('renderRoundedNumber', () => {
  test('returns the desired placeholder value if passed a null', () =>
    expect(renderRoundedNumber(null)).toEqual('â€”'))

  test('trims zeros when necessary', () =>
    expect(renderRoundedNumber(1.2, { precision: 15 })).toMatchSnapshot())

  test('uses proper highlight color', () =>
    expect(renderRoundedNumber(1.2, { highlightColor: '#ABCDEF' })).toMatchSnapshot())

  test('allows for arbitrary formatting of tooltip', () =>
    expect(
      renderRoundedNumber(1.2, {
        formatTooltip: (rounded) =>
          `I AM A ROUND NUMBER AS FOLLOWS: "${rounded}" THANK YOU FOR YOUR TIME`,
      })
    ).toMatchSnapshot())
})
