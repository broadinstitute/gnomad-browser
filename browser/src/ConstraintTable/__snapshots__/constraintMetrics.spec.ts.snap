// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderRoundedNumber allows for arbitrary formatting of tooltip 1`] = `
<TooltipAnchor
  tooltip="I AM A ROUND NUMBER AS FOLLOWS: "1.200" THANK YOU FOR YOUR TIME"
  tooltipComponent={[Function]}
>
  <TooltipHint>
    1.2
  </TooltipHint>
</TooltipAnchor>
`;

exports[`renderRoundedNumber trims zeros when necessary 1`] = `
<TooltipAnchor
  tooltip="1.200"
  tooltipComponent={[Function]}
>
  <TooltipHint>
    1.2
  </TooltipHint>
</TooltipAnchor>
`;

exports[`renderRoundedNumber uses proper highlight color 1`] = `
<TooltipAnchor
  tooltip="1.200"
  tooltipComponent={[Function]}
>
  <constraintMetrics__ConstraintHighlight
    highlightColor="#ABCDEF"
  >
    1.2
  </constraintMetrics__ConstraintHighlight>
</TooltipAnchor>
`;
